### PR TITLE
Disambiguate bins

### PR DESCRIPTION
**Pull request by lukpueh**
## This PR does
- above all try to disambiguate *targets* and delegated targets roles, i.e. *bins* and *bin-n* roles,
- it further fleshes out the *roles* entry in the *definitions* sections and removes the obsolete clarification why *snapshot* is not called *release*,
- it also fixes spelling.

## This PR does NOT (but should) solve the following issues:

- Why doesn't targets delegate directly to *bin-n* roles? What is the benefit of having an intermediate *bins* role? The only purpose I can see is semantic. Delegating from *targets* directly to *bin-n* would reduce complexity of this pep.

- ~~The recommended number of bins was adjusted to 16,384, we should adopt that number in Figure 2 that shows `bin 1` to `bin 1024`, alternatively (and preferably) we can replace `1024` with `N`.~~

- In section "Revoking Trust in Projects and Versions":
  - What is the difference between revoking trust in projects and revoking it in versions? If I understand correctly in both cases, the target files corresponding to that project or version need to be removed from the *bin-n* metadata they are listed in, which need to be re-signed with the online key. If that's accurate, why differentiate?
  - Also, is there a case where the *bins* role (offline key) needs to be resigned?

- In section "Key Compromise Analysis":
  - Disambiguate *bins* and *bin-n* in table.
  - Clarify what "projects targets metadata" is (paragraph under the table).
  - Maybe re-think the table. There are multiple instances that describe how,
 snapshot, timestamp and/or bin-n roles need to cooperate for an attacker to be successful, which seems less meaningful when these roles share the same online key, which is recommend in this pep.

Please review and and above all **comment** to the concerns and questions I am raising here!

 (cc @JustinCappos, @trishankatdatadog, @mnm678, @SantiagoTorres, @awwad. I'm also inviting @joshuagl who kindly volunteered to help.)

### CODE REVIEW

**Comment by JustinCappos** [comment]
> Why doesn't targets delegate directly to bin-n roles? What is the benefit of having an intermediate bins role? The only purpose I can see is semantic. Delegating from targets directly to bin-n would reduce complexity of this pep.

It also makes it easier to revoke a single key or change the top level targets (to transition to PEP 480).  You don't need to re-download the bins -> bin-n mapping in this case, which is fairly large.

> What is the difference between revoking trust in projects and revoking it in versions? If I understand correctly in both cases, the target files corresponding to that project or version need to be removed from the bin-n metadata they are listed in, which need to be re-signed with the online key. If that's accurate, why differentiate?

It's the same in PEP 458.  They are different in PEP 480, which was what I was thinking of with my original PR, so this should be fixed.

> Also, is there a case where the bins role (offline key) needs to be resigned?

Not unless there was a weakness in how the keys were generated.

File: pep-0458.txt [hunk 1]
```diff
@@ -166,8 +166,9 @@ Terms used in this PEP are defined as follows:
 * Roles: There is one *root* role in PyPI.  There are multiple roles whose
   responsibilities are delegated to them directly or indirectly by the *root*
   role. The term top-level role refers to the *root* role and any role
-  delegated by the *root* role. Each role has a single metadata file that it is
-  trusted to provide.
+  delegated directly by the *root* role, i.e. *timestamp*, *snapshot* and
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
  specified directly by the *root* role, i.e. *timestamp*, *snapshot* and
```

Comment (trishankatdatadog) [reply pep-0458.txt]: Let's not overload the term "delegate"

File: pep-0458.txt [hunk 2]
```diff
@@ -340,15 +338,15 @@ delegated to by *bins* (i.e., *bin-n*).  The *timestamp* and *snapshot*
 metadata MUST be updated whenever *root*, *targets* or delegated metadata are
 updated.  Observe, though, that *root* and *targets* metadata are much less
 likely to be updated as often as delegated metadata.  Similarly, the *bins* role
-will only be updated when new bins are added.  Therefore, *timestamp*,
+will only be updated when a new *bin-n* role is added.  Therefore, *timestamp*,
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
will only be updated when a *bin-n* role is added, updated, or removed.  Therefore, *timestamp*,
```

File: pep-0458.txt [hunk 3]
```diff
@@ -640,12 +638,11 @@ include a version number in their filename:
...
     where VERSION_NUMBER is an incrementing integer, and ROLENAME is one of the
     top-level metadata roles -- *root*, *snapshot* or *targets* -- or one of
-    the delegated targets roles -- *bins* or *bin <i>*.
+    the delegated targets roles -- *bins* or *bin-n*.
 
 The only exception is the *timestamp* metadata file, whose version is not known
-in advance, when a client performs an update.
-
-Instead the *timestamp* metadata serves as version root. That is, it lists the
+in advance, when a client performs an update. Instead the *timestamp* metadata
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
in advance, when a client performs an update. The *timestamp* metadata
```

File: pep-0458.txt [hunk 4]
```diff
@@ -640,12 +638,11 @@ include a version number in their filename:
...
+serves as version root. That is, it lists the
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
lists the
```

File: pep-0458.txt [hunk 5]
```diff
@@ -640,12 +638,11 @@ include a version number in their filename:
...
 version of the *snapshot* metadata, which in turn lists the versions of *root*,
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
version of the *snapshot* metadata, which in turn lists the versions of the
```

File: pep-0458.txt [hunk 6]
```diff
@@ -685,26 +682,24 @@ efficiently transfer consistent snapshots from PyPI.
...
+on how PyPI will respond to a project transaction.
 
 When a project uploads a new transaction, the project transaction process MUST
-add all new targets and relevant delegated *bins* metadata.  (It is shown later
-in this section why the *bins* role will delegate targets to a number of
-delegated *bins* roles.)  Finally, the project transaction process MUST inform
-the snapshot process about new delegated *bins* metadata.
+add all new targets and relevant delegated *bin-n* metadata. Finally, the
+project transaction process MUST inform the snapshot process about new
+delegated *bin-n* metadata.
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
*bin-n* metadata.
```

File: pep-0458.txt [hunk 7]
```diff
@@ -685,26 +682,24 @@ efficiently transfer consistent snapshots from PyPI.
...
+a project transaction is discussed in a later section.  For now, the focus is
+on how PyPI will respond to a project transaction.
 
 When a project uploads a new transaction, the project transaction process MUST
-add all new targets and relevant delegated *bins* metadata.  (It is shown later
-in this section why the *bins* role will delegate targets to a number of
-delegated *bins* roles.)  Finally, the project transaction process MUST inform
-the snapshot process about new delegated *bins* metadata.
+add all new targets and relevant delegated *bin-n* metadata. Finally, the
+project transaction process MUST inform the snapshot process about new
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
project transaction process MUST inform the snapshot process about any new
```

File: pep-0458.txt [hunk 8]
```diff
@@ -685,26 +682,24 @@ efficiently transfer consistent snapshots from PyPI.
...
+delegated *bin-n* metadata.
 
 Project transaction processes SHOULD be automated and MUST also be applied
 atomically: either all metadata and targets -- or none of them -- are added.
 The project transaction and snapshot processes SHOULD work concurrently.
-Finally, project transaction processes SHOULD keep in memory the latest *bins*
+Finally, project transaction processes SHOULD keep in memory the latest *bin-n*
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
Finally, project transaction processes SHOULD use the latest *bin-n*
```

File: pep-0458.txt [hunk 9]
```diff
@@ -728,7 +722,8 @@ Snapshot Process
 
 The snapshot process is fairly simple and SHOULD be automated.  The snapshot
 process MUST keep in memory the latest working set of *root*, *targets*, and
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
process SHOULD use the latest working set of the *targets* and all
```

File: pep-0458.txt [hunk 10]
```diff
@@ -728,7 +722,8 @@ Snapshot Process
...
-delegated roles.  Upon an update, the snapshot process will sign for this
+delegated targets roles (i.e. *bins* and *bin-n* roles).  Upon an update, the
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
delegated targets roles' (i.e. *bins* and *bin-n* roles) metadata.  Upon an update, the
```

**Review by trishankatdatadog — CHANGES_REQUESTED** [review]
VERY good observations, Lukas, thanks! I just have some minor changes, please.

I'll try to work separately on your other comments...

File: pep-0458.txt [hunk 11]
```diff
@@ -685,26 +682,24 @@ efficiently transfer consistent snapshots from PyPI.
...
 
 Project transaction processes SHOULD be automated and MUST also be applied
 atomically: either all metadata and targets -- or none of them -- are added.
 The project transaction and snapshot processes SHOULD work concurrently.
-Finally, project transaction processes SHOULD keep in memory the latest *bins*
+Finally, project transaction processes SHOULD keep in memory the latest *bin-n*
 metadata so that they will be correctly updated in new consistent snapshots.
 
-Signing updated snapshot, timestamp, and bin metadata needs to be done on each 
+Signing updated *snapshot*, *timestamp*, and *bin-n* metadata needs to be done on each
```
Comment (trishankatdatadog) [inline pep-0458.txt]:
```suggestion
Signing updated *timestamp*, *snapshot*, and *bin-n* metadata needs to be done on each
```

**Comment by trishankatdatadog** [comment]
@lukpueh pushed your other comments [here](https://github.com/secure-systems-lab/peps/issues/15)

**Comment by lukpueh** [comment]
Thanks for the updates @trishankatdatadog! I batch committed them in c8b0657. Let's merge here and address the remaining comments in #15 (FYI, I also moved @JustinCappos comment over there).

**Pull request merged**